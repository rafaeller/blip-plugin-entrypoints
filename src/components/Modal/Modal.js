import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import styles from './modal.module.scss';

import { ReactComponent as CloseIcon } from '../../assets/images/svg/close-icon.svg';
import CustomButton from '../CustomButton';

const Modal = ({
    title,
    children,
    footer,
    width = 50,
    closable = true,
    visible = false,
    hasFooter = false,
    elementRef = null,
    confirmLabel = 'Salvar',
    cancelLabel = 'Cancelar',
    onConfirm = () => {},
    onCancel = () => {},
    onClose = () => {}
}) => {
    const [isVisible, setIsVisible] = useState(visible);
    const [customPosition, setCustomPosition] = useState({});

    const wraper_class = classNames(
        'h-100 w-100 center pa0 pt4-ns fixed z-9999 dn',
        {
            [styles['modal-open']]: !!isVisible
        }
    );

    const mask_class = classNames(
        'h-100 w-100 absolute top-0 left-0 bg-black-70',
        styles['modal-mask']
    );

    const content_class = classNames(
        'w-100 relative center',
        styles['modal-content'],
        {
            [`w-${width}-ns`]: !!width
        }
    );

    useEffect(() => {
        if (elementRef) {
            const { offsetTop } = elementRef.current;
            setCustomPosition({ paddingTop: offsetTop });
        } else {
            setCustomPosition({});
        }
    }, [elementRef]);

    useEffect(() => {
        setIsVisible(!!visible);
    }, [visible]);

    const handleClose = () => {
        setIsVisible(false);
        onClose();
    };

    const renderHeader = () => (
        <div className="relative flex items-center">
            {!!title && (
                <div className="f3 fw6 lh-copy ma0 mw6 truncate bp-c-neutral-dark-desk">
                    {title}
                </div>
            )}
            {!!closable && (
                <div
                    aria-label="Close"
                    data-testid="close-modal-icon"
                    className="pointer ml-auto bp-c-neutral-dark-rooftop"
                    onClick={() => handleClose()}
                >
                    <CloseIcon />
                </div>
            )}
        </div>
    );

    const renderFooter = () => (
        <div className="flex justify-end">
            <div className="mr2">
                <CustomButton
                    text={cancelLabel}
                    variant="secondary"
                    onClick={() => {
                        onCancel();
                        handleClose();
                    }}
                    dataTestId="cancel-modal-button"
                />
            </div>
            <bds-button
                variant='primary'
                onClick={() => onConfirm()}
                dataTestId="confirm-modal-button"
            >
                {confirmLabel}
            </bds-button>
        </div>
    );

    return (
        <div className={wraper_class} style={customPosition}>
            <div
                className={mask_class}
                data-testid="close-overlay"
                onClick={() => handleClose()}
            ></div>
            <div className={content_class}>
                <div className="bg-white shadow-6 br3 pa4">
                    <div className="h-100 w-100 flex flex-column flex-nowrap">
                        {(!!title || !!closable) && renderHeader()}
                        <div className="pt1-5 pb1-5">{children}</div>
                        {hasFooter && (footer || renderFooter())}
                    </div>
                </div>
            </div>
        </div>
    );
};

Modal.propTypes = {
    title: PropTypes.string,
    children: PropTypes.any,
    footer: PropTypes.any,
    width: PropTypes.number,
    closable: PropTypes.bool,
    visible: PropTypes.bool,
    hasFooter: PropTypes.bool,
    elementRef: PropTypes.any,
    confirmLabel: PropTypes.string,
    cancelLabel: PropTypes.string,
    onConfirm: PropTypes.func,
    onCancel: PropTypes.func,
    onClose: PropTypes.func
};

export default Modal;
