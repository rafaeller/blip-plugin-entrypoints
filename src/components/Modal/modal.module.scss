.modal-open {
    display: block;
}

.modal-open .modal-mask, .modal-open .modal-content {
    animation: FadeIn 0.3s cubic-bezier(0.7, 0.3, 0.1, 1);
    opacity: 1;
    pointer-events: auto;
    transition: none;
}

.modal-mask, .modal-content {
    opacity: 0;
    pointer-events: auto;
    transition: opacity 0.3s ease;
}

@keyframes FadeIn {
    0% {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
}