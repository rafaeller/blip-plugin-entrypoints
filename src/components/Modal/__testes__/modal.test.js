import React from 'react';
import '@testing-library/jest-dom/extend-expect';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Modal from '../Modal';

describe('Modal component', () => {
    const mockClose = jest.fn();
    const mockCancel = jest.fn();
    const mockConfirm = jest.fn();

    beforeEach(() =>
        render(
            <Modal
                title="My Modal"
                visible={true}
                hasFooter={true}
                onConfirm={mockConfirm}
                onCancel={mockCancel}
                onClose={mockClose}
            >
                Content
            </Modal>
        )
    );

    afterEach(() => jest.clearAllMocks());

    it('should handle cancel click event', () => {
        const buttonNode = screen.queryByTestId('cancel-modal-button');
        userEvent.click(buttonNode);
        expect(mockCancel).toHaveBeenCalledTimes(1);
    });

    it('should handle confirm click event', () => {
        const buttonNode = screen.queryByTestId('confirm-modal-button');
        userEvent.click(buttonNode);
        expect(mockConfirm).toHaveBeenCalledTimes(1);
    });

    it('should handle close modal when click on overlay', () => {
        const node = screen.queryByTestId('close-overlay');
        userEvent.click(node);
        expect(mockClose).toHaveBeenCalledTimes(1);
    });

    it('should handle close modal when click on close icon', () => {
        const node = screen.queryByTestId('close-modal-icon');
        userEvent.click(node);
        expect(mockClose).toHaveBeenCalledTimes(1);
    });

    it('should render correctly', () => {
        const { container } = render(
            <Modal title="My Modal" visible={true}>
                Content
            </Modal>
        );
        expect(container.firstChild).toMatchSnapshot();
    });
});
