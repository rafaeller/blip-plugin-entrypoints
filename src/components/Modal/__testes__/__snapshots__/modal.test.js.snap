// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should render correctly 1`] = `
<div
  class="h-100 w-100 center pa0 pt4-ns fixed z-9999 dn modal-open"
>
  <div
    class="h-100 w-100 absolute top-0 left-0 bg-black-70 modal-mask"
    data-testid="close-overlay"
  />
  <div
    class="w-100 relative center modal-content w-50-ns"
  >
    <div
      class="bg-white shadow-6 br3 pa4"
    >
      <div
        class="h-100 w-100 flex flex-column flex-nowrap"
      >
        <div
          class="relative flex items-center"
        >
          <div
            class="f3 fw6 lh-copy ma0 mw6 truncate bp-c-neutral-dark-desk"
          >
            My Modal
          </div>
          <div
            aria-label="Close"
            class="pointer ml-auto bp-c-neutral-dark-rooftop"
            data-testid="close-modal-icon"
          >
            <svg>
              close-icon.svg
            </svg>
          </div>
        </div>
        <div
          class="pt1-5 pb1-5"
        >
          Content
        </div>
      </div>
    </div>
  </div>
</div>
`;
